src/main/java/com/puppycrawl/tools/checkstyle/xpath/XpathQueryGenerator.java

```diff
@@ -0,0 +1,229 @@
+     */
+    private static List<DetailAST> getMatchingAstElements(DetailAST root, int lineNumber,
+                                                          int columnNumber) {
+        final List<DetailAST> result = new ArrayList<>();
+        DetailAST curNode = root;
+        while (curNode != null && curNode.getLineNo() <= lineNumber) {
+            if (curNode.getLineNo() == lineNumber && curNode.getColumnNo() == columnNumber
+                    && curNode.getType() != TokenTypes.IDENT) {
+                result.add(curNode);
+            }
+            DetailAST toVisit = curNode.getFirstChild();
+            while (curNode != null && toVisit == null) {
+                toVisit = curNode.getNextSibling();
+                if (toVisit == null) {
+                    curNode = curNode.getParent();
+                }
+            }
+
+            curNode = toVisit;
+        }
+        return result;
+    }
+
+    /**
+     * Returns full xpath query for given ast element.
+     * @param root {@code DetailAST} root element
+     * @param ast {@code DetailAST} ast element
+     * @return full xpath query for given ast element
+     */
+    private static String getXpathQuery(DetailAST root, DetailAST ast) {
+        final StringBuilder result = new StringBuilder("");
+        DetailAST cur = ast;
+        while (cur != root) {
+            final StringBuilder curNodeQueryBuilder = new StringBuilder("/")
+                    .append(TokenUtils.getTokenName(cur.getType()));
+            final DetailAST identAst = cur.findFirstToken(TokenTypes.IDENT);
+            if (identAst != null) {
+                curNodeQueryBuilder.append("[@text='")
+                        .append(identAst.getText())
+                        .append("']");
+            }
+            result.insert(0, curNodeQueryBuilder);
+            cur = cur.getParent();
+        }
+        return result.toString();
+    }
+
+    /**
+     * To be sure that generated xpath query will return exactly required ast element, the element
+     * should be checked for uniqueness. If ast element has {@link TokenTypes#IDENT} as the child
+     * or there is no sibling with the same {@code TokenTypes} then element is supposed to be
+     * unique. This method finds if {@code DetailAst} element is unique.
+     * @param ast {@code DetailAST} ast element
+     * @return if {@code DetailAst} element is unique
+     */
+    private static boolean isUniqueAst(DetailAST ast) {
+        boolean isOnly = true;
+        final DetailAST identAst = ast.findFirstToken(TokenTypes.IDENT);
+
+        if (identAst == null) {
+            DetailAST cur;
+            if (ast.getParent() == null) {
+                cur = ast;
+                while (cur.getPreviousSibling() != null) {
+                    cur = cur.getPreviousSibling();
+                }
+            }
+            else {
+                cur = ast.getParent().getFirstChild();
+            }
+
+            while (cur != null) {
```

Comment:
Here you check the siblings. So you can extract the code to the method and give it the user-friendly name to increase the readability of your algorithm.